### PR TITLE
Shared: Inline test expectations

### PR DESCRIPTION
**Pull request by yoff**
Moves the inline test implementation to a shared module.

### CODE REVIEW

**Comment by erik-krogh** [comment]
I had an idea of how to simplify the signature even further: We could make a `newtype` based implementation of `Location` inside the `Make` module, and by doing that just use a `hasLocationInfo` predicate instead of `getLocation()`.  

I've implemented it as a PR against this PR here: https://github.com/yoff/codeql/pull/37

File: cpp/ql/test/TestUtilities/InlineExpectationsTest.qll [hunk 1]
```diff
@@ -1,399 +1,35 @@
...
- *
- * is equivalent to:
- *
- * `tag1=expected-value tag2=expected-value`
+ * Inline expectation tests for C++.
+ * See `shared/util/codeql/util/test/InlineExpectationsTest.qll`
  */
 
-private import InlineExpectationsTestPrivate
+import cpp as C
```
Comment (jketema) [inline cpp/ql/test/TestUtilities/InlineExpectationsTest.qll:6]: Is there any particular reason a named import is being used?

**Review by jketema — COMMENTED** [review]
When expanding some of the individual files in the "Files changed" view I still see some QL-for-QL warnings, e.g., on line 16 of the new `cpp/ql/test/TestUtilities/InlineExpectationsTest.qll`. Those should probably be resolved.

File: java/ql/test/TestUtilities/InlineExpectationsTest.qll [hunk 2]
```diff
@@ -1,399 +1,46 @@
...
+    override string getContents() { result = this.getText().suffix(2).trim() }
   }
-}
 
-/* Note: These next three classes correspond to all the possible values of type `TColumn`. */
-class GoodExpectation extends ValidExpectation {
-  GoodExpectation() { getKnownFailure() = "" }
-}
+  private class XmlExpectationComment extends ExpectationComment instanceof J::XmlComment {
+    override string getContents() { result = this.(J::XmlComment).getText().trim() }
```
Comment (atorralba) [inline java/ql/test/TestUtilities/InlineExpectationsTest.qll:32]:
```suggestion
    override string getContents() { result = super.getText().trim() }
```

File: java/ql/test/TestUtilities/InlineExpectationsTest.qll [hunk 3]
```diff
@@ -1,399 +1,46 @@
...
 
-class FalsePositiveExpectation extends ValidExpectation {
-  FalsePositiveExpectation() { getKnownFailure() = "SPURIOUS" }
-}
+    override Location getLocation() { result = this.(J::XmlComment).getLocation() }
```
Comment (atorralba) [inline java/ql/test/TestUtilities/InlineExpectationsTest.qll:34]:
```suggestion
    override Location getLocation() { result = J::XmlComment.super.getLocation() }
```

File: java/ql/test/TestUtilities/InlineExpectationsTest.qll [hunk 4]
```diff
@@ -1,399 +1,46 @@
...
 
-class FalseNegativeExpectation extends ValidExpectation {
-  FalseNegativeExpectation() { getKnownFailure() = "MISSING" }
-}
+    override predicate hasLocationInfo(string path, int sl, int sc, int el, int ec) {
+      this.(J::XmlComment).hasLocationInfo(path, sl, sc, el, ec)
```
Comment (atorralba) [inline java/ql/test/TestUtilities/InlineExpectationsTest.qll:37]:
```suggestion
      J::XmlComment.super.hasLocationInfo(path, sl, sc, el, ec)
```

File: java/ql/test/TestUtilities/InlineExpectationsTest.qll [hunk 5]
```diff
@@ -1,399 +1,46 @@
...
+    }
 
-class InvalidExpectation extends Expectation, TInvalidExpectation {
-  string expectation;
-
-  InvalidExpectation() { this = TInvalidExpectation(comment, expectation) }
+    override string toString() { result = this.(J::XmlComment).toString() }
```
Comment (atorralba) [inline java/ql/test/TestUtilities/InlineExpectationsTest.qll:40]:
```suggestion
      J::XmlComment.super.hasLocationInfo(path, sl, sc, el, ec)
```

**Review by atorralba — COMMENTED** [review]
Java mostly LGTM, with a small suggestion:

**Review by aibaars — DISMISSED** [review]
:+1: For Ruby

**Comment by yoff** [comment]
@erik-krogh nice, I agree we might prefer that..

File: config/identical-files.json [hunk 6]
```diff
@@ -402,16 +402,6 @@
...
-  "Inline Test Expectations": [
-    "cpp/ql/test/TestUtilities/InlineExpectationsTest.qll",
-    "csharp/ql/test/TestUtilities/InlineExpectationsTest.qll",
-    "java/ql/test/TestUtilities/InlineExpectationsTest.qll",
-    "python/ql/test/TestUtilities/InlineExpectationsTest.qll",
-    "ruby/ql/test/TestUtilities/InlineExpectationsTest.qll",
-    "ql/ql/test/TestUtilities/InlineExpectationsTest.qll",
-    "go/ql/test/TestUtilities/InlineExpectationsTest.qll",
-    "swift/ql/test/TestUtilities/InlineExpectationsTest.qll"
-  ],
```
Comment (geoffw0) [inline config/identical-files.json:414]: 🎉

**Review by geoffw0 — COMMENTED** [review]
:+1: for CPP and Swift, if the warnings pointed out by @jketema are addressed.

File: config/identical-files.json (hunk 6 above)
Comment (erik-krogh) [reply config/identical-files.json:414]: Friday win for sure.

**Comment by yoff** [comment]
Unfortunately we _do_ have to assume a `Location` class, otherwise users cannot implement `hasActualResult`. (And I think that changing the signature of that predicate to not use `Location` would be cumbersome.)

**Review by erik-krogh — APPROVED** [review]
👍

File: go/ql/test/TestUtilities/InlineExpectationsTest.qll [hunk 7]
```diff
@@ -1,399 +1,22 @@
...
 
+private module Impl implements InlineExpectationsTestSig {
   /**
-   * Holds if there is an optional result on the specified location.
-   *
-   * This is similar to `hasActualResult`, but returns results that do not require a matching annotation.
-   * A failure will still arise if there is an annotation that does not match any results, but not vice versa.
-   * Override this predicate to specify optional results.
+   * Represents a line comment in the Go style.
+   * include the preceding comment marker (`//`).
```
Comment (owen-mc) [inline go/ql/test/TestUtilities/InlineExpectationsTest.qll:12]:
```suggestion
   * A class representing line comments n the Go style, including the preceding
   * comment marker (`//`).
```

**Review by owen-mc — APPROVED** [review]

**Review by jketema — DISMISSED** [review]

**Review by jketema — APPROVED** [review]

**Review by owen-mc — APPROVED** [review]

**Review by atorralba — APPROVED** [review]
Java 👍

**Review by erik-krogh — APPROVED** [review]

**Review by michaelnebel — APPROVED** [review]
C# LGTM!
Very nice application of param modules!!!

**Review by tausbn — APPROVED** [review]
Python 👍

**Comment by hvitved** [comment]
OOI: Why didn't we replace the `abstract string` class `InlineExpectationsTest` with a parameterized module as well?

**Comment by erik-krogh** [comment]
> OOI: Why didn't we replace the `abstract string` class `InlineExpectationsTest` with a parameterized module as well?

That would require rewriting lots of the tests that use inline-expectation-tests.   
This PR was just made to share more code.  

But it sounds like a nice idea to rewrite `InlineExpectationsTest` to a parameterized module.

**Pull request merged**